Write out a best_epoch file and reform callbacks

davidwyld opened the pull request:
Here I've introduced a few lines into the evaluate callback to update the best epoch and keep a symlink named `best_epoch.h5` up to date. This will allow a user to just refer to a model's best epoch in a standard form for prediction.

This is a draft PR at the moment, it would be good if you guys can check this approach and let me know if you have any feedback.

Something that needs fixing is that at the moment this just checks that the metric is biggest, but this obviously only works for metrics like Pearson Correlation Coefficient that need to be maximised. Another metric like RMSE or MAE may need to be minimised (correct me if I'm wrong @fredblain ). Fred, could you also give some feedback as to whether these are the main metrics used or whether we might need to account for something else? In this case, the metrics are set in qe_metrics, so we have control over that.

Before merging, I would also like to tidy up the callbacks in general. I've noticed that some of the callbacks specified in deepquest.utils.callbacks aren't actually used (my fault, I don't think I fully thought about what was happening when I moved them there). Also now, predict is working a bit differently on the `hotfix-predict` branch, so will try to account for that (or just fix the relevant callbacks over on that branch.

Commit:
symlink the best epoch file to best_epoch.h5

fredblain commented on deepquest/utils/callbacks.py:
```diff
@@ -477,6 +479,17 @@ def evaluate(self, epoch, counter_name='epoch', logs={}):
+                dst = self.save_path + '/' + 'best_' + counter_name + '.h5'
+                os.symlink(src, tmp)
+                os.rename(tmp, dst) # gets around overwriting/updating symlinks
+
```
This is not necessary as the test whether the current epoch improves over the past one is done lines 1191-1224 in the `evaluate()` function of the `EarlyStopping` class:
```python
        # Check if the best score has been outperformed in the current epoch
        if current_score > self.best_score:
            self.best_epoch = epoch
            self.best_score = current_score
            self.threshold = threshold
            self.wait = 0
            if self.verbose > 0:
                if self.threshold is not None:
                    logging.info('---current best %s %s: %.3f threshold %.1f' % (self.check_split, self.metric_check,
                                                              current_score if not self.want_to_minimize
                                                              else -current_score, self.threshold))
                else:
                    logging.info('---current best %s %s: %.3f' % (self.check_split, self.metric_check,
                                                              current_score if not self.want_to_minimize
                                                              else -current_score))

        # Stop training if performance has not improved for self.patience epochs
        elif self.patience > 0:
            self.wait += 1
            logging.info('---bad counter: %d/%d' % (self.wait, self.patience))
            if self.wait >= self.patience:
                if self.verbose > 0:
                    if self.threshold is not None:
                        logging.info("---%s %d: early stopping. Best %s found at %s %d: %f threshold %.1f" % (
                             str(counter_name), epoch, self.metric_check, str(counter_name), self.best_epoch,
                             self.best_score if not self.want_to_minimize else -self.best_score, self.threshold))
                    else:
                        logging.info("---%s %d: early stopping. Best %s found at %s %d: %f" % (
                             str(counter_name), epoch, self.metric_check, str(counter_name), self.best_epoch,
                             self.best_score if not self.want_to_minimize else -self.best_score))

                self.model.stop_training = True
                exit(1)
```
The block above is what we see at the end of each training epoch.
My understanding is that we simply have to save the model here under `best_model.h5`, overwriting the previous one if it exists. Check the use of the `saveModel()` function for that, as in the `evaluate()` function of `EvalPerformance` (lines 475-478):
```python
        # Save the model
        if self.save_each_evaluation:
            from keras_wrapper.cnn_model import saveModel
            saveModel(self.model_to_eval, epoch, store_iter=not self.eval_on_epochs)
```
One issue I see with this function is that it doesn't take a "model name", therefore we need to find a workaround here. We could have our own local version of the function, but we depend a lot on `Keras_wrapper` therefore I would like to avoid copying a few functions here and there into deepQuest's codebase, such that if they change something one day (e.g. like having the architecture and the weights in a single file -- something they did when we were looking into that with @drj11 at the very begining) it shouldn't break and our goal is to keep the maintenance load as low as possible on that aspect.

Also, I would not use simlink as well, because if the parameter `SAVE_EACH_EVALUATION` is set to `False`, the file `epoch_X.h5` is not created. Same if the parameter `EPOCHS_FOR_SAVE` is bigger than 1.

davidwyld replied:
Perhaps I should have explained, sorry, the `EarlyStopping` callback used in training is specified in`keras_wrapper.cnn_model` (line 991, in `Model_Wrapper.__train`), imported on line 27 of the same file `from keras_wrapper.extra.callbacks import *` so we can't edit that callback. Hence why I put the code in with our `deepquest.utils.callbacks.evaluation()`
Correct me if I'm wrong or you have a solution.

To confuse matters, there's a version of EarlyStopping in `deepquest.utils.callbacks`, which is unused and I want to get rid of.

I will change the symlink to a model save.

davidwyld replied:
Have changed to using `saveModel` in 50cc2e6, certainly as it is currently written, we can hijack use the options of saveModel to give it a full path to save as follows:
```python
best_path = self.save_path + '/' + 'best_' + counter_name
from keras_wrapper.cnn_model import saveModel
saveModel(self.model_to_eval, epoch, store_iter=False, path=best_path, full_path=True)
```
which seems to do the trick.

fredblain replied:
OK, that's good that we can use `path` to set the name. 
However, using `best_ + counter_name` will result in filling the disk space unnecessarily as it would simply duplicate each epoch until no improvement is observed. We should use `best_model.h5` as a "checkpoint" instead, overwriting the previous one if it exits.

davidwyld replied:
That is what happens here, the value of `counter_name` is (confusingly) the string `'epoch'` (not a digit), so the file written out goes to `trained_modes/model_name/best_epoch.h5` which naturally gets overwritten every time there's a better epoch.

fredblain replied:
OK, my bad!

davidwyld replied:
No wukkas. I'll ask @drj11 if he thinks this approach is okay and we can merge?

fredblain replied:
Can we make sure it works whatever the considered metrics or level of prediction, before we merge? I can see there is still a FIXME for instance.

fredblain replied:
In the callback blocks that have been removed, there is a mechanism you may want to reuse:

```python
        if self.want_to_minimize:
            current_score = -current_score
```

davidwyld replied:
Yup, sorry got a bit over excited. Yes, this is probably the way I was going to do it. I see that there are a number of different metrics in `deepquest.utils.evaluation`:
- pearson, rmse and mae for sentences (& documents)
- Precision, Recall, f1_prod, and threshold for words
- Threshold, Precision, Recall, F-score, and F-score multi for phrases

`want_to_minimize` is slightly troubling in that it is only set when calling `EarlyStopping` during `Model_wrapper.__train()` and currently as:
```python
# Early stopper
if params.get('metric_check') is not None:
    callback_early_stop = EarlyStopping(self,
                                        patience=params['patience'],
                                        metric_check=params['metric_check'],
                                        want_to_minimize=True if 'TER' in params['metric_check'] else False,
                                        min_delta=params['min_delta'],
                                        eval_on_epochs=params['eval_on_epochs'],
                                        each_n_epochs=params['each_n_epochs'],
                                        start_eval_on_epoch=params['start_eval_on_epoch'])
    callbacks.append(callback_early_stop)
```
and so at present, is only set to True if the metric is `'TER'` (weirdly this is the same in the old codebase) so perhaps was only run with metrics to be maximised?

One way around this is to prevent `__train()` from adding its own instance of the EarlyStopping callback and add our own instance, setting `want_to_minimize` on our own terms.

fredblain replied:
RMSE and MAE are both error metrics hence to be minimised, although the metric of interest for training sent- and doc-level models is Pearson's r correlation (to be maximised).

davidwyld replied:
This is added by c2278fa

fredblain reviewed (dismissed):
I commented it on below.

Commit:
change best model output to use keras_wrapper.cnn_model.saveModel

Commit:
remove redundant callbacks

drj11 commented:
It should be a thought in the back of our minds that at some point prediction will want to an ensemble of, say, the top 3 models or something similar.

drj11 commented on deepquest/train.py:
```diff
@@ -14,7 +14,7 @@
 
 import deepquest.qe_models as modFactory
-from deepquest.utils.callbacks import PrintPerformanceMetricOnEpochEndOrEachNUpdates
+from deepquest.utils.callbacks import PrintPerformanceMetricOnEpochEndOrEachNUpdates#, EarlyStopping
```
that seems untidy

davidwyld replied:
Thanks for the spot, this is left over from some testing. Will go through and clean it up before merging.

Commit:
facilitate training on minimised metrics (ie rmse and mae)

this commit makes changes to the way the EarlyStopping callback from Keras Wrapper
is added to the callbacks for training. It is now added in deepquest.train.buildCallbacks
which allows us to set the want_to_minimize parameter dependent on our metrics
previously is was hard-set in keras_wrapper.extra.callbacks and was in effect, a bug.

Further, this commit removes the model save from deepquest.callbacks.EvalPerformance.evaluation
as saveModel is already called by keras_wrapper's on_epoch_end callback which was resulting
in the model being saved twice each epoch.

davidwyld commented on deepquest/utils/callbacks.py:
```diff
@@ -474,15 +476,11 @@ def evaluate(self, epoch, counter_name='epoch', logs={}):
-        if self.save_each_evaluation:
-            from keras_wrapper.cnn_model import saveModel
-            saveModel(self.model_to_eval, epoch, store_iter=not self.eval_on_epochs)
-
```
This commit removes the model save here since it is already saved by `on_epoch_end` which was resulting in the model being saved twice every epoch. @fredblain are you aware why this may be incorrect or why we may need extra parameters to account for this kind of behaviour, there are a lot of parameters..

davidwyld commented:
@fredblain I've made some changes in commit c2278fa to fix a bug where 'rmse' and 'mae' metrics were not actually minimised in training and also facilitated the correct checking of these for the 'best_model' feature.
